Assets/MixedRealityToolkit-SDK/Features/Input/MixedRealityInputManager.cs

```diff
@@ -1702,5 +1743,156 @@ public void RaiseDictationError(IMixedRealityInputSource source, string dictatio
+            {
+                var digitalRules = CurrentInputActionRulesProfile.InputActionRulesSingleAxis;
+
+                for (int i = 0; i < digitalRules.Length; i++)
+                {
+                    if (digitalRules[i].BaseAction == inputAction && digitalRules[i].Criteria.Equals(criteria))
+                    {
+                        if (digitalRules[i].RuleAction == inputAction)
+                        {
+                            Debug.LogError("Input Action Rule cannot be the same as the rule's Base Action!");
+                            return inputAction;
+                        }
+
+                        return digitalRules[i].RuleAction;
+                    }
+                }
+            }
+
+            return inputAction;
+        }
+
+        private MixedRealityInputAction ProcessRules(MixedRealityInputAction inputAction, Vector2 criteria)
+        {
+            if (CurrentInputActionRulesProfile != null)
+            {
+                var digitalRules = CurrentInputActionRulesProfile.InputActionRulesDualAxis;
+
+                for (int i = 0; i < digitalRules.Length; i++)
+                {
+                    if (digitalRules[i].BaseAction == inputAction)
+                    {
+                        if (digitalRules[i].Criteria == criteria)
+                        {
+                            if (digitalRules[i].RuleAction == inputAction)
+                            {
+                                Debug.LogError("Input Action Rule cannot be the same as the rule's Base Action!");
+                                return inputAction;
+                            }
+
+                            return digitalRules[i].RuleAction;
+                        }
+                    }
+                }
+            }
+
+            return inputAction;
+        }
+
+        private MixedRealityInputAction ProcessRules(MixedRealityInputAction inputAction, Vector3 criteria)
+        {
+            if (CurrentInputActionRulesProfile != null)
+            {
+                var digitalRules = CurrentInputActionRulesProfile.InputActionRulesVectorAxis;
+
+                for (int i = 0; i < digitalRules.Length; i++)
+                {
+                    if (digitalRules[i].BaseAction == inputAction && digitalRules[i].Criteria == criteria)
+                    {
+                        if (digitalRules[i].RuleAction == inputAction)
+                        {
+                            Debug.LogError("Input Action Rule cannot be the same as the rule's Base Action!");
+                            return inputAction;
+                        }
+
+                        return digitalRules[i].RuleAction;
+                    }
+                }
+            }
+
+            return inputAction;
+        }
+
+        private MixedRealityInputAction ProcessRules(MixedRealityInputAction inputAction, Quaternion criteria)
+        {
+            if (CurrentInputActionRulesProfile != null)
```

Comment:
A lot of this is the same code for each type. Possibly there is a way to genericise it, maybe?

Reply:
I tried, but unfortunately none that was immediately obvious. If you figure out a way I'm happy to accept a PR